helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/constraints/ConstraintBasedAlgorithm.java

```diff
@@ -152,69 +169,135 @@ private double getAssignmentNormalizedScore(AssignableNode node, AssignableRepli
-            // the smaller priority number means higher priority.
-            return statePriority1 - statePriority2;
-          }
+  private Map<String, Integer> computeOverallClusterRemainingCapacity(List<AssignableNode> nodes) {
+    Map<String, Integer> utilizationMap = new HashMap<>();
+    for (AssignableNode node : nodes) {
+      for (String capacityKey : node.getMaxCapacity().keySet()) {
+        if (utilizationMap.containsKey(capacityKey)) {
+          Integer newUtil =
+              utilizationMap.get(capacityKey) + node.getRemainingCapacity().get(capacityKey);
+          // update util
+          utilizationMap.put(capacityKey, newUtil);
         } else {
-          // If the baseline assignment contains the assignment, prioritize the replica.
-          return baselineAssignment.containsKey(resourceName1) ? -1 : 1;
+          Integer newUtil = node.getRemainingCapacity().get(capacityKey);
+          utilizationMap.put(capacityKey, newUtil);
         }
+      }
+    }
+    return utilizationMap;
+  }
+
+  /**
+   * Update the overallClusterUtilMap with newly placed replica
+   */
+  private void updateOverallClusterRemainingCapacity(
+      Map<String, Integer> overallClusterRemainingCapacityMap, AssignableReplica replica) {
+    for (Map.Entry<String, Integer> resourceUsage : replica.getCapacity().entrySet()) {
+      Integer newRemainingCapacity =
+          overallClusterRemainingCapacityMap.get(resourceUsage.getKey()) - resourceUsage.getValue();
+      overallClusterRemainingCapacityMap.put(resourceUsage.getKey(), newRemainingCapacity);
+    }
+  }
+
+  private class AssignableReplicaWithScore implements Comparable<AssignableReplicaWithScore> {
+    private final AssignableReplica _replica;
+    private int _score = 0;
+    private boolean _isInBestPossibleAssignment;
+    private boolean _isInBaselineAssignment;
+    private final Integer  _replicaHash;
+
+    AssignableReplicaWithScore(AssignableReplica replica, ClusterModel clusterModel) {
+      _replica = replica;
+      _isInBestPossibleAssignment = clusterModel.getContext().getBestPossibleAssignment()
+          .containsKey(replica.getResourceName());
+      _isInBaselineAssignment =
+          clusterModel.getContext().getBaselineAssignment().containsKey(replica.getResourceName());
+      _replicaHash = Objects.hash(replica.toString(), clusterModel.getAssignableNodes().keySet());
+    }
+
+    public void reComputeScore(Map<String, Integer> overallClusterRemainingCapMap) {
+      int score = 0;
+      // score = SUM(weight * (resource_capacity/cluster_capacity) where weight = 1/(1-total_util%)
+      // it could be be simplified to "resource_capacity/cluster_remainingCapacity".
+      for (Map.Entry<String, Integer> resourceCapacity : _replica.getCapacity().entrySet()) {
+        score += resourceCapacity.getValue() * 100 / (
+            overallClusterRemainingCapMap.get(resourceCapacity.getKey())
+                + 0.01f /*avoid divide by 0*/);
```

Comment:
Does 0.01 really make sense, in case when a certain capacity is measured in decimals, and 0.01 becomes a significant portion (total cluster capacity = 0.01 TB, as an extreme example, then adding 0.01 is a 100% increase to the remaining cap). 

On the other hand, why are we handling the case for "remaining cap = 0"? When that happens, this assignment is already doomed, right? I don't think it's possible to place this replica any more.

Reply:
Thanks for the comment. I think it reveals a valid point -- should we continue the process when remaining capacity is 0? I think we could safe several compute cycle by checking hard constrain first and quit early. That might be more elegant comparing to this hacky way of avoiding divide by 0. 
Regarding the example in your comment, I am not sure if this will happen since the remaining capacity is an Integer. I think adding 0.01 to 1 might not change the result a lot since we only care about the relative relationship between each replicas. 
Considering the effort in development, I think maybe adding 0.01 here is the easiest way to avoid divide by 0. 

All comments are welcome.  :D

Reply:
My original comment is more focused on the fact that 0.01 is a magical number that could have an impact to the remaining capacities, since the remaining capacities aren't normalized and their units can vary. You're right that it's an integer so my extreme example wouldn't hold, but I hope my point has come across. :) 

I firmly believe that using a magical number to avoid "divide by 0" is hacky no matter what value is chosen. I suggest throwing an exception when the remaining capacity is 0 and terminate the assignment right then and there. For code cleanness, you may want to consider using do while as suggested by @jiajunwang . I will also comment there to see if that solution makes sense.

Reply:
We had an offline discussion with jiajunwang. Our final decision is:
1. function `computeScore` should add 0 to final score when replica's capacity on a dimension is 0.
2. function `computeScore` will return Integer.MAX_VALUE when cluster's remaining capacity is 0.
3. `computeScore` won't throw exception or do admission check (won't throw exception when cluster's remaining capacity is 0). Constrain check will fail this placement in the next step.
4. Also, to avoid overflow, the score is caped at 100 for each dimension.